### PR TITLE
[Manure] Updates Contributing Factors to Anaerobic Lagoon Methane Calculations

### LINKED ISSUE #2630: [Manure] Update anaerobic lagoon methane emission methods
Context: We have received feedback from Nicole that the lack of difference between emissions from slurry storages and anaerobic lagoons has caused some confusion among FARM ES end users. Additionally, due to management, climate, and differing manure conditions in storage (e.g., temperature, volume to depth ratio, loading rate, retention, microbial population composition and activity, and other factors), methane emissions from anaerobic lagoons are generally considered to be greater than from stored slurry manure. With this, the manure module team has sought guidance from external SMEs on what parameters we should consider modifying, and are ready to try out several solutions.

- [x] **Task 1: Create a testing branch**
- The dev who takes on this issue needs to first create a testing branch with the following changes in order for @elle-andreen and other manure SMEs to evaluate the exact changes that will ultimately be implemented. The testing branch will have two major new features. 
- Feature 1 is the ability to modulate the lower bound of manure temperature in an anaerobic lagoon. To do this, create a constant for `minimum_anaerobic_lagoon_temperature` with a value of 5 (C). In `_determine_outdoor_storage_temperature` in `processor.py`, utilize this constant in place of the current lower bound of 0.0. 
- Feature 2 is the ability to retain a set proportion of manure in anaerobic lagoons at emptying (i.e. reaching end of storage time). The % of manure retained should be a constant, which also must be created (you can name it `anaerobic_lagoon_manure_retention` or similar and set to 0.25 to starch). For each variable in ManureStream, at emptying, instead of resetting all values to 0.0, reset values to "`anaerobic_lagoon_manure_retention` x accumulated value in storage". This is meant to represent removing 75% of manure in the lagoon when storage time is reached, as opposed to 100%. @elle-andreen can provide more detail on this as needed.

- [x] **Task 2: Implement changes into dev.**
- When SME testing on the testing branch is complete and desired minimum temperature and retention values have been identified, implement the features described above with their finalized constant values into dev and close this issue.

### PR DESCRIPTION
**Pull request by ew3361zh**
### Context

Issue(s) closed by this pull request: closes #2630

### What

Adds 3 constants:
1. `ANAEROBIC_LAGOON_MINIMUM_TEMPERATURE` - new lower bound temp for anaerobic lagoon manure.
2. `SLURRY_OUTDOOR_MINIMUM_TEMPERATURE` - new lower bound temp for SSO manure.
3. `ANAEROBIC_LAGOON_MANURE_RETENTION` - new retention proportion for anaerobic lagoon storages implemented on emptying days.

- Updates `METHANE_TO_METHANE_CARBON_DIOXIDE_RATIO` to `VS_TO_METHANE_LOSS_RATIO`.

### Why

Updates contributing factors to AnLag methane production calculations to be closer to expected scientific outputs.

Note about name change for `VS_TO_METHANE_LOSS_RATIO` constant:

The value of METHANE_TO_METHANE_CARBON_DIOXIDE_RATIO was updated from 9.25 to 6.665, and its name was modified to VS_TO_METHANE_LOSS_RATIO, to correct an error in interpretation of the quantity of VS lost per unit of CH4 emitted in [Petersen et al., 2015](https://www.sciencedirect.com/science/article/pii/S0301479724002196). Initially, based on the method description in this paper, CH4 and CO2 were assumed to be lost in a molar ratio of 1:3 (on a mass basis, for every mole of CH4 emitted, 3 moles of CO2 were lost). As CO2 has a mass 8.25x greater than CH4, the ratio value was originally set to 9.25 (1 + 1x8.25), and all VS loss was assumed to come from CO2 and CH4 emission (e.g. VS loss = CH4 + CO2 loss). However, in the [source material](http://dce2.au.dk/pub/SR437.pdf) Petersen et al. 2015 references for this ratio, it is explained more clearly that the 1:3 ratio refers specifically to moles of C lost. On other words, for each C lost as CH4, 3 C are assumed to be lost as CO2. Additionally, manure VS are assumed to be 45% C. Therefore, the mass ratio of total VS loss to CH4 emitted is 6.665. A brief mathematical explanation of this is below. The name of the constant was updated to more accurately reflect its value, as it now relates directly to VS loss, not the ratio of CH4 to CO2 emission.
Screenshot 2025-12-01 114253

### How

1. To implement `MINIMUM_ANAEROBIC_LAGOON_TEMPERATURE` and `SLURRY_OUTDOOR_MINIMUM_TEMPERATURE` constants - added new arg `minimum_manure_temperature` to `Storage._determine_outdoor_storage_temperature()`. In both the `SSO` and `AnaerobicLagoon` calls to this function, it sends this lower-bound constant to overwrite the default. This function is only used by outdoor manure storages.
2. To implement `ANAEROBIC_LAGOON_MANURE_RETENTION` constant - added new property `_emptying_fraction()` at the `Storage` level which defaults to `1.0` matching behavior on dev for all storages. Gave `AnaerobicLagoon` an overwriting property version which defaults to the new constant. A property seemed like the best way to potentially allow this to be more extendible/flexible in the future if the fraction needs to be calculated or applied differently by different `Storage` child classes. Within the parent `Storage.process_manure()` where `emptying_day` is checked for and the storages are emptied, I implemented a check of the value for this `emptying_fraction` and split the manure accordingly.
3. Updated all references of `METHANE_TO_METHANE_CARBON_DIOXIDE_RATIO` to `VS_TO_METHANE_LOSS_RATIO`. 

### Test plan

Testing was run on the `example_open_lot_task.json` task because that task uses AnaerobicLagoon storage. The splitting was occurring as expected to the best of my knowledge and the new lower bound temp was enacted as expected as well.

### Input Changes

N/A

### Output Changes

- CH4 emissions from slurry storages are slightly increased due to the change in the quantity of VS destroyed per unit of CH4 emitted, i.e., more VS are retained in storage and thus subsequent emissions are elevated
- NH3 emissions are slightly increased in anaerobic lagoons, largely due to the implementation of default retention at emptying
- CH4 emissions per unit of VS loaded into storage (i.e. methane conversion factors) are increased in anaerobic lagoons, due to the combined effect of the implemented minimum temperature and retention. Values now align better with IPCC emission estimates; emissions are overall higher and the differential between cold and warm climates is increased. Changes in MCF compared to dev (including recent temperature modeling updates) across a few weather files:
image

#### Filter

```json
Manure.Storage.AnaerobicLagoon.*
Manure.Storage.SlurryStorageOutdoor.*
```

### CODE REVIEW

File: input/task_manager_metadata.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
     "tasks": {
       "title": "Task manager data",
       "description": "Configuration file for general simulation parameters.",
-      "path": "input/data/tasks/example_freestall_task.json",
+      "path": "input/data/tasks/example_open_lot_task.json",
```
Comment (ew3361zh) [inline input/task_manager_metadata.json]: will change this back before merging.

**Comment by ew3361zh** [comment]
After conversation with @elle-andreen and @matthew7838, the lower-bound temperature changes were removed from this PR for the time being to see what the effects will be from #2636 first because the lower-bound temp changes might not be necessary after that's implemented.

File: RUFAS/biophysical/manure/storage/storage.py [hunk 2]
```diff
@@ -136,9 +144,20 @@ def process_manure(self, _: CurrentDayConditions, time: RufasTime) -> dict[str,
...
+                self.stored_manure = retained_stream
+                manure_to_be_returned = {"manure": replace(emptied_stream)}
+            elif self._emptying_fraction <= 0.0:
+                empty_stream = ManureStream.make_empty_manure_stream()
+                self._report_manure_stream(empty_stream, "emptied", time.simulation_day)
+                manure_to_be_returned = {}
+            else:
+                self._report_manure_stream(self.stored_manure, "emptied", time.simulation_day)
+                manure_to_be_returned = {"manure": replace(self.stored_manure)}
+                self.stored_manure = ManureStream.make_empty_manure_stream()
```
Comment (allisterakun) [inline RUFAS/biophysical/manure/storage/storage.py:215]: unit test?

**Review by allisterakun — APPROVED** [review]
LGTM!

File: RUFAS/biophysical/manure/storage/storage.py (hunk 2 above)
Comment (ew3361zh) [reply RUFAS/biophysical/manure/storage/storage.py:215]: Will add - was waiting until we had scientific approval to do the final touches but won't merge in before that's added.

File: RUFAS/biophysical/manure/storage/anaerobic_lagoon.py [hunk 3]
```diff
@@ -48,6 +48,11 @@ def __init__(
             capacity=capacity,
         )
 
+    @property
+    def _emptying_fraction(self) -> float:
+        """The fraction of manure to be emptied from the Anaerobic Lagoon when it is time to empty it."""
+        return 1.0 - ManureConstants.ANAEROBIC_LAGOON_MANURE_RETENTION
```
Comment (matthew7838) [inline RUFAS/biophysical/manure/storage/anaerobic_lagoon.py:56]: We should also have some safety checks here to make sure the fraction always falls into the range between 0 and 1. Again, not a problem here, but it could be a potential problem if someone decided to modify the constants.

**Review by matthew7838 — APPROVED** [review]
LGTM. Left a few comments regarding potential invalid values.

File: RUFAS/biophysical/manure/storage/storage.py (hunk 2 above)
Comment (matthew7838) [inline RUFAS/biophysical/manure/storage/storage.py:215]: I think there should be some error handling at the calculation of this "fraction" or here to prevent it from being > 1 or < 0. From my understanding, I believe this shouldn't be a problem in this PR, but I don't think we want to leave the possibility open.

Comment (ew3361zh) [reply RUFAS/biophysical/manure/storage/storage.py:215]: The logic here is:
1. 0 < `_emptying_fraction` < 1.0 -> empty the amount based on the fraction and retain the rest.
2. `_emptying_fraction` <= 0.0 -> report an empty stream and return an empty stream.
3. `_emptying_fraction` >= 1.0 -> empty the full stream and return the original `stored_stream`.

So we're handling those instances where the fraction is outside the 0.0-1.0 range by either emptying the whole storage or not emptying it at all.

Not sure if that's "too helpful" and covering up a user overwriting a constant with a number they'd want to break the simulation or if it's fine but maybe we should just be more transparent about what's happening with their bad fraction there.

@elle-andreen @KFosterReed what are your thoughts here? Do you like handling the cases where `Storage`.`process_manure()` receives an `_emptying_fraction` outside the 0-1.0 range in the way described above or would you prefer to log a bad fraction and stop the simulation or maybe just add a warning to let the user know they've created an `emptying_fraction` outside the bounds but we're proceeding by either emptying the the full amount if the fraction is 1.0 or greater or emptying none of it if the fraction is 0.0 or less?

File: RUFAS/biophysical/manure/storage/anaerobic_lagoon.py (hunk 3 above)
Comment (ew3361zh) [reply RUFAS/biophysical/manure/storage/anaerobic_lagoon.py:56]: Please see my other comment response to your other error handling feedback - I'm not sure what we'd like to happen here. My instinct is that maybe adding a warning is fine and we could "handle" their changes to the constant in the ways it's currently handled or add warnings that their changes led to a bad `_emptying_fraction` or stop the simulation altogether.

File: RUFAS/biophysical/manure/storage/storage.py (hunk 2 above)
Comment (elle-andreen) [reply RUFAS/biophysical/manure/storage/storage.py:215]: @ew3361zh @matthew7838 @KFosterReed Thank you for the thoughtful questions on this! My inclination here is to error out anything above 1 or below 0. Niko has a good point that we're effectively handling those cases from a code standpoint, e.g. not creating literal mathematical problems by emptying negative or >100 percents, but I think it would still be helpful. This is one of the parameters in the manure module that I'd consider somewhat more likely to be modified by folks working directly with the code, and I think there's no harm in ensuring folks do not unknowingly enter 50 instead of 0.50 and not have an error returned

**Comment by ew3361zh** [comment]
@elle-andreen @KFosterReed @matthew7838 updates to this PR:

- changed the default `ANAEROBIC_LAGOON_MANURE_RETENTION` constant from `0.25` -> `0.1`
- added a validation method for the storage `_emptying_fraction` that will raise an error if it's outside `0.0`-`1.0` range.
- changed `manure_to_be_returned` to original `self.stored_manure` for manure processing in storage sub-classes when splitting retained/emptied manure.
- updated associated unit testing

We've got 2 approvals already but if we could get a SME go-ahead to merge, that would be helpful to verify these late additions are looking as hoped!

**Comment by KFosterReed** [comment]
I tested the Manure Tem updates in PR #2636 and then pulled those into this branch to test the combined effects and overall, I think it is an improvement in that it widened the differences in MCF in different climates and between slurry storage and anareobic lagoons. Here is a summary of my results. 

image

One question I have is if there is still a plan to implement a minimum temperature for the anaerobic lagoon? With the temperature updates, I was still getting manure temperatures below 0 deg C in the cold climate scenario I tested.

**Comment by elle-andreen** [comment]
> One question I have is if there is still a plan to implement a minimum temperature for the anaerobic lagoon? With the temperature updates, I was still getting manure temperatures below 0 deg C in the cold climate scenario I tested.

@KFosterReed Yes! We will still implement the 1 C minimum temperature, the intention is that the temperature modeling logic in Matthew's PR is used but any modeled values < 1 C are are just equal to 1 (e.g. lower bound of 1, not "recalibrating" the temperature curve to never dip below 1). Because Matthew's PR is modifying the same function where that lower bound will be implemented, Niko is (to the best of my knowledge) just waiting for Matthew's to be finalized before he makes that change.

**Comment by matthew7838** [comment]
> > One question I have is if there is still a plan to implement a minimum temperature for the anaerobic lagoon? With the temperature updates, I was still getting manure temperatures below 0 deg C in the cold climate scenario I tested.
> 
> @KFosterReed Yes! We will still implement the 1 C minimum temperature, the intention is that the temperature modeling logic in Matthew's PR is used but any modeled values < 1 C are are just equal to 1 (e.g. lower bound of 1, not "recalibrating" the temperature curve to never dip below 1). Because Matthew's PR is modifying the same function where that lower bound will be implemented, Niko is (to the best of my knowledge) just waiting for Matthew's to be finalized before he makes that change.

@elle-andreen Just to confirm, should the lower bound be implemented on my PR or this one?

**Comment by elle-andreen** [comment]
> > > One question I have is if there is still a plan to implement a minimum temperature for the anaerobic lagoon? With the temperature updates, I was still getting manure temperatures below 0 deg C in the cold climate scenario I tested.
> > 
> > 
> > @KFosterReed Yes! We will still implement the 1 C minimum temperature, the intention is that the temperature modeling logic in Matthew's PR is used but any modeled values < 1 C are are just equal to 1 (e.g. lower bound of 1, not "recalibrating" the temperature curve to never dip below 1). Because Matthew's PR is modifying the same function where that lower bound will be implemented, Niko is (to the best of my knowledge) just waiting for Matthew's to be finalized before he makes that change.
> 
> @elle-andreen Just to confirm, should the lower bound be implemented on my PR or this one?

@matthew7838 I believe on Niko's PR, since that change is specific to just anaerobic lagoons.

File: RUFAS/biophysical/manure/storage/storage.py [hunk 4]
```diff
@@ -104,9 +104,18 @@ def is_overflowing(self) -> bool:
         """True if the manure in storage exceeds the storage's volumetric capacity, else False."""
         return self.stored_manure.volume > self._capacity
 
+    @property
+    def _emptying_fraction(self) -> float:
+        """
+        The fraction of the storage that is emptied when the storage is emptied. Defaults to 1.0
```
Comment (elle-andreen) [inline RUFAS/biophysical/manure/storage/storage.py]:
```suggestion
        The fraction of the accumulated stored manure that is removed from storage when the emptying time is reached. Defaults to 1.0
```

File: RUFAS/biophysical/manure/storage/anaerobic_lagoon.py [hunk 5]
```diff
@@ -48,6 +48,11 @@ def __init__(
             capacity=capacity,
         )
 
+    @property
+    def _emptying_fraction(self) -> float:
+        """The fraction of manure to be emptied from the Anaerobic Lagoon when it is time to empty it."""
```
Comment (elle-andreen) [inline RUFAS/biophysical/manure/storage/anaerobic_lagoon.py]:
```suggestion
        """The fraction of the accumulated stored manure that is removed from storage when the emptying time is reached."""
```

**Review by elle-andreen — APPROVED** [review]
All appears to be working here as intended! Thanks for your work on this Niko!
